src/coreclr/jit/codegencommon.cpp

```diff
@@ -2805,1407 +2805,568 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
-                    if (insertArg)
-                    {
-                        assert(!regArgTab[currentArgNum].processed);
-
-                        // EA_4BYTE is probably wrong here (and below)
-                        // todo -- suppress self move
-                        GetEmitter()->emitIns_R_R_I_I(INS_mov, EA_4BYTE, destRegNum, regNum,
-                                                      regArgTab[currentArgNum].slot - 1, 0);
-                        assert(!genIsValidIntReg(destRegNum) || !genIsValidFloatReg(regNum));
-                        regArgTab[currentArgNum].processed = true;
-                        regArgMaskLive &= ~genRegMask(regNum);
-                    }
-                }
+                sourceReg = graph->GetOrAdd(REG_NEXT(sourceReg->reg));
+                destReg   = graph->GetOrAdd(REG_NEXT(destReg->reg));
+                graph->AddEdge(sourceReg, destReg, TYP_FLOAT, 0);
+                continue;
             }
+#endif
+            graph->AddEdge(sourceReg, destReg, edgeType, seg.Offset - baseOffset);
         }
     }
-#endif // defined(TARGET_ARM64) && defined(FEATURE_SIMD)
+}
 
-    /* Finally take care of the remaining arguments that must be enregistered */
-    while (regArgMaskLive)
+// -----------------------------------------------------------------------------
+// genHomeRegisterParams: Move all register parameters to their initial
+// assigned location.
+//
+// Parameters:
+//    initReg            - A register that this method should communicate if it becomes non-zero
+//    initRegStillZeroed - [out] whether or not initReg is still zeroed
+//
+void CodeGen::genHomeRegisterParams(regNumber initReg, bool* initRegStillZeroed)
+{
+#ifdef DEBUG
+    if (verbose)
     {
-        regMaskTP regArgMaskLiveSave = regArgMaskLive;
+        printf("*************** In genHomeRegisterParams()\n");
+    }
+#endif
 
-        for (argNum = 0; argNum < argMax; argNum++)
+    regMaskTP paramRegs = intRegState.rsCalleeRegArgMaskLiveIn | floatRegState.rsCalleeRegArgMaskLiveIn;
+    if (compiler->opts.OptimizationDisabled())
+    {
+        // All registers are going to frame
+        for (unsigned lclNum = 0; lclNum < compiler->info.compArgsCount; lclNum++)
         {
-            /* If already processed go to the next one */
-            if (regArgTab[argNum].processed)
+            LclVarDsc* lclDsc = compiler->lvaGetDesc(lclNum);
+
+            if (!lclDsc->lvOnFrame)
             {
                 continue;
             }
 
-            if (regArgTab[argNum].slot == 0)
-            { // Not a register argument
-                continue;
+            const ABIPassingInformation& abiInfo = compiler->lvaParameterPassingInfo[lclNum];
+            for (unsigned i = 0; i < abiInfo.NumSegments; i++)
+            {
+                const ABIPassingSegment& seg = abiInfo.Segments[i];
+                if (seg.IsPassedInRegister() && ((paramRegs & genRegMask(seg.GetRegister())) != 0))
```

Comment:
I guess I'm confused by the second part of the `&&` here -- what are the cases where we won't (or don't need) to store?

Reply:
It's basically the liveness check, but LSRA uses a check that's a bit different when creating intervals and marking live registers in this set:

https://github.com/dotnet/runtime/blob/8e2655b932f5f5e184d289982bdc516576df0981/src/coreclr/jit/lsrabuild.cpp#L2329-L2338

Not totally convinced that LSRA and this code shouldn't be able to use more precise `bbLiveIn`, but I guess that's something for another day.

Reply:
I suppose this particular one you're commenting on has to be based on `lvRefCnt()` since it is for MinOpts. But the one inside `genSpillOrAddRegisterParam` could probably use real liveness instead.